### DISCUSSION TITLE
How to extend lib

### DISCUSSION THREAD

**Discussion by Lykos153** [hercules-ci/flake-parts#234] [Q&A]
Hey :) 

I have a flake with various `perSystem` modules in their own files. Many of them need access to a custom implementation of [mkRenamedOptionModule](https://noogle.dev/f/lib/modules/mkRenamedOptionModule). The only way I found to do this is:
```nix
{lib, ...}: {
  mylib = import ../../../nix/module/lib.nix {inherit lib;};
  imports = [ (mylib.mkRenamedOptionModule ["a"] ["b"])];
}
```
Which is kind of ugly because in each file I need to get the relative path to `lib.nix` right and moving them is a nightmare.

I came across `config._module.args.mylib = ...` in https://github.com/hercules-ci/flake-parts/discussions/222 but I can't use that in `imports` because that would cause an infinite recursion.

So what I'm looking for is a way to add my function to `lib`, eg. at `lib.mylib.mkRenamedOptionModule` so that all modules have access to them by default. How can I achieve this?

**Comment by srid** [comment]
FWIW, I did some exploration on this [here](https://nixos.zulipchat.com/#narrow/stream/419910-flake-parts/topic/Overriding.20.60lib.60.20in.20flake-parts).

**Comment by roberth** [comment]
Hello Silvio,
Nothing is impossible, but I don't recommend literally extending `lib` because _as a general pattern_ it's not good for reusability. If you or others do this often, you're effectively creating one or many little dialects that prevents reuse of your modules, without much visual indication in the module that there could be problems.
So what I would recommend instead is to use `specialArgs` (it's a `mkFlake` argument too) to put your own library under a recognizable name that isn't `lib`.
